pylhc/irnl_rdt_correction.py

```diff
@@ -0,0 +1,1124 @@
+            _check_corrector_order_not_lower(rdt, correctors)
+            _check_update_optics(rdt, correctors, rdt_map, update_optics, feed_down)
+
+
+def _check_update_optics(rdt: RDT, correctors: list, rdt_map: dict, update_optics: bool, feed_down: int):
+    """ Check if corrector values are set before they would be needed for feed-down. """
+    if not update_optics:
+        return
+
+    for corrector in correctors:
+        corrector_order = int(corrector[1])
+        for rdt_comp in rdt_map.keys():
+            if rdt_comp.order == rdt.order:
+                break
+            if rdt_comp.order <= corrector_order <= rdt_comp.order + feed_down:
+                raise ValueError(
+                    "Updating the optics is in this configuration not possible,"
+                    f" as corrector {corrector} influences {rdt_comp.name}"
+                    f" with the given feeddown of {feed_down}. Yet the value of"
+                    f" the corrector is defined by {rdt.name}.")
+
+
+def _check_corrector_order_not_lower(rdt, correctors):
+    """ Check if only higher and equal order correctors are defined to correct
+    a given rdt."""
+    wrong_correctors = [c for c in correctors if int(c[1]) < rdt.order]
+    if len(wrong_correctors):
+        raise ValueError(
+            "Correctors can not correct RDTs of higher order."
+            f" Yet for {rdt.name} the corrector(s)"
+            f" '{list2str(wrong_correctors)}' was (were) given."
+        )
+
+
+# IO Handling ------------------------------------------------------------------
+# Input ---
+
+def _check_opt(opt: DotDict) -> DotDict:
+    # Set defaults
+    for name, default in DEFAULTS.items():
+        if opt.get(name) is None:
+            LOG.debug(f"Setting input '{name}' to default value '{default}'.")
+            opt[name] = default
+
+    # check accel
+    opt.accel = opt.accel.lower()  # let's not care about case
+    if opt.accel not in DEFAULT_RDTS.keys():
+        raise ValueError(f"Parameter 'accel' needs to be one of '{list2str(list(DEFAULT_RDTS.keys()))}' "
+                         f"but was '{opt.accel}' instead.")
+
+    # Set rdts:
+    if opt.get('rdts') is None:
+        opt.rdts = DEFAULT_RDTS[opt.accel]
+
+    # Check required and rdts:
+    for name in ('optics', 'errors', 'beams', 'rdts'):
+        inputs = opt.get(name)
+        if inputs is None or isinstance(inputs, str) or not isinstance(inputs, (Iterable, Sized)):
+            raise ValueError(f"Parameter '{name}' is required and needs to be "
+                             "iterable, even if only of length 1. "
+                             f"Instead was '{inputs}'.")
+
+    if opt.feeddown < 0 or not (opt.feeddown == int(opt.feeddown)):
+        raise ValueError("'feeddown' needs to be a positive integer.")
+
+    if opt.iterations < 1:
+        raise ValueError("At least one iteration (see: 'iterations') needs to "
+                         "be done for correction.")
+    return opt
```

Comment:
what if beam 1,2 & 4 are given? shouldn't 2&4 be mutually exclusive?

Reply:
It actually doesn't matter what beams are given. You can also give it two beam 1 optics, if you want to optimize for two different optics at the same time for some reason.